client/src/main/java/org/eclipse/hono/client/impl/GatewayMappingCommandHandler.java

```diff
@@ -0,0 +1,97 @@
+/*******************************************************************************
+ * Copyright (c) 2019 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+
+package org.eclipse.hono.client.impl;
+
+import org.apache.qpid.proton.amqp.transport.ErrorCondition;
+import org.eclipse.hono.client.Command;
+import org.eclipse.hono.client.CommandContext;
+import org.eclipse.hono.client.GatewayMapper;
+import org.eclipse.hono.tracing.TracingHelper;
+import org.eclipse.hono.util.CommandConstants;
+import org.eclipse.hono.util.Constants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+
+/**
+ * Command handler that maps commands to gateway devices (if applicable) and passes the commands to a given
+ * next command handler.
+ */
+public class GatewayMappingCommandHandler implements Handler<CommandContext> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(GatewayMappingCommandHandler.class);
+
+    private final GatewayMapper gatewayMapper;
+    private final Handler<CommandContext> nextCommandHandler;
+
+    /**
+     * Creates a new GatewayMappingCommandHandler instance.
+     * 
+     * @param gatewayMapper The component mapping a command device id to the corresponding gateway device id.
+     * @param nextCommandHandler The handler to invoke with the mapped command.
+     */
+    public GatewayMappingCommandHandler(final GatewayMapper gatewayMapper, final Handler<CommandContext> nextCommandHandler) {
+        this.gatewayMapper = gatewayMapper;
+        this.nextCommandHandler = nextCommandHandler;
+    }
+
+    @Override
+    public void handle(final CommandContext originalCommandContext) {
+        final Command originalCommand = originalCommandContext.getCommand();
+        if (!originalCommand.isValid()) {
+            originalCommandContext.reject(new ErrorCondition(Constants.AMQP_BAD_REQUEST, "malformed command message"));
```

Comment:
do we still need to execute the rest of the method if the command is being rejected?
I think we also should re-issue some credit so that the application can send another command, don't you think?

Reply:
Oh, thanks for spotting, missed the `return` statement there.
As for the credits: The `TenantScopedCommandConsumer` has its receiver link being created with the default (ie. non-empty) `prefetchSize`, so the credit handling should be automatic (in contrast to the `DeviceSpecificCommandConsumer` with `prefetchSize=0`).